openmdao/components/tests/test_exec_comp.py

```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
         self.assertTrue("balance" in text)
         self.assertTrue("x_1" in text)
 
+
+class TestFunctionRegistration(unittest.TestCase):
+
+    # These 2 tests don't run normally unless you run testflo with a -m "featuretest_*"
+    # I don't want to include them in our normal tests because they don't contain
+    # the _temporary_expr_dict contextmanager (to avoid user confusion)
+    # which keeps the contents of the ExecComp _expr_dict clean
+    def featuretest_register_simple(self):
+        import openmdao.api as om
+
+        om.ExecComp.register("myfunc", lambda x: x * x, complex_safe=True)
+        p = om.Problem()
+        comp = p.model.add_subsystem("comp", om.ExecComp("y = 2 * myfunc(x)"))
+
+        p.setup()
+        p.run_model()
+        J = p.compute_totals(of=['comp.y'], wrt=['comp.x'])
+        assert_near_equal(J['comp.y', 'comp.x'][0][0], 4., 1e-10)
+
+    def featuretest_register_simple_unsafe(self):
+        import openmdao.api as om
+
+        # the following function isn't really complex unsafe, but we'll call it unsafe anyway
+        # for demonstration purposes
+        om.ExecComp.register("unsafe", lambda x: x * x, complex_safe=False)
+        p = om.Problem()
+        comp = p.model.add_subsystem("comp", om.ExecComp("y = 2 * unsafe(x)"))
+
+        # because our function is complex unsafe, we must declare that the partials
+        # with respect to 'x' use 'fd' instead of 'cs'
+        comp.declare_partials('*', 'x', method='fd')
+
+        p.setup()
+        p.run_model()
+        J = p.compute_totals(of=['comp.y'], wrt=['comp.x'])
+        assert_near_equal(J['comp.y', 'comp.x'][0][0], 4., 1e-5)
+
+    def test_register_simple(self):
+        with _temporary_expr_dict():
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)'))
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            self.assertEqual(p['comp.area_square'], 9.)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            self.assertEqual(J['comp.area_square', 'comp.x'], 6.)
+
+    def test_register_simple_arr(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            p.setup()
+            p['comp.x'] = 3.
```

Comment:
Using the same value for all of the vector inputs might hide a problem.

Reply:
fixed